### PR TITLE
 Add static type to host functions

### LINKED ISSUE #1136: Add static types to HostFunctionValue
### Issue To Be Solved
Currently, host functions don't have static types

### Suggested Solution

Pass the static type during construction of host function value and store it in value itself.

### PR DESCRIPTION
**Pull request by SupunS**
Closes #1136

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/interpreter.go [hunk 1]
```diff
@@ -2175,50 +2188,56 @@ func (interpreter *Interpreter) functionConditionsWrapper(
...
+					preConditions,
+					body,
+					rewrittenPostConditions,
+					returnType,
+				)
+			},
+
+			// This is an internally created and used function, and can
+			// never pass around as a value. Hence, type is not required.
+			nil,
```
Comment (turbolent) [inline runtime/interpreter/interpreter.go:2239]: 👍

File: runtime/interpreter/interpreter.go [hunk 2]
```diff
@@ -2175,50 +2188,56 @@ func (interpreter *Interpreter) functionConditionsWrapper(
...
+					beforeStatements,
+					preConditions,
+					body,
+					rewrittenPostConditions,
+					returnType,
+				)
+			},
+
+			// This is an internally created and used function, and can
+			// never pass around as a value. Hence, type is not required.
```
Comment (turbolent) [inline runtime/interpreter/interpreter.go:2238]:
```suggestion
			// never be passed around as a value. Hence, the type is not required.
```

File: runtime/interpreter/interpreter.go [hunk 3]
```diff
@@ -2580,6 +2599,10 @@ var converterFunctionValues = func() []converterFunction {
 			func(invocation Invocation) Value {
 				return convert(invocation.Arguments[0])
 			},
+
+			// Converter functions are not pass around as values.
+			// Hence, type is not required.
```
Comment (turbolent) [inline runtime/interpreter/interpreter.go:2604]:
```suggestion
			// Converter functions are not passed around as values.
			// Hence, the type is not required.
```

File: runtime/tests/interpreter/container_mutation_test.go [hunk 4]
```diff
@@ -258,6 +259,104 @@ func TestArrayMutation(t *testing.T) {
...
+				},
+				Options: []interpreter.Option{
+					interpreter.WithPredeclaredValues(values),
+				},
+			},
+		)
+
+		require.NoError(t, err)
+
+		// TODO: Shouldn't throw an error once dynamic subtyping for functions is implemented.
```
Comment (turbolent) [inline runtime/tests/interpreter/container_mutation_test.go:294]: Could you please create a follow-up issue for this TODO?

**Review by turbolent — APPROVED** [review]
Great work! 👏 

It might be worth it to cache the result of the `ConvertStaticToSemaType` in the values to avoid convering them each time. Maybe best to do this in a follow-up PR

File: runtime/tests/interpreter/container_mutation_test.go (hunk 4 above)
Comment (SupunS) [reply runtime/tests/interpreter/container_mutation_test.go:294]: Created: https://github.com/onflow/cadence/issues/1145

**Comment by SupunS** [comment]
> It might be worth it to cache the result of the ConvertStaticToSemaType in the values to avoid converting the static types to sema types each time. Maybe best to do this in a follow-up PR

For function types, it shouldn't be too much of an overhead, since it only creates a thin wrapper. But definitely, we could cache these, and also cache parameter types, return types, etc in the `FunctionStaticType`.

**Pull request merged**